woudc_data_registry/dataset_validators.py

```diff
@@ -0,0 +1,640 @@
+
+import logging
+import statistics
+
+from collections import OrderedDict
+
+
+LOGGER = logging.getLogger(__name__)
+
+DATASETS = ['Broad-band', 'Lidar', 'Multi-band', 'OzoneSonde', 'RocketSonde',
+            'Spectral', 'TotalOzone', 'TotalOzoneObs', 'UmkehrN14']
+
+
+def get_validator(dataset):
+    """
+    Returns a DatasetValidator instance tied to <dataset>.
+    If <dataset> is a valid data category but no special validator exists
+    for it, returns a base validator that automatically succeeds.
+
+    :param dataset: Name of a WOUDC data type.
+    :returns: Validator class targetted to that data type.
+    """
+
+    if dataset == 'TotalOzone':
+        return TotalOzoneValidator()
+    elif dataset == 'TotalOzoneObs':
+        return TotalOzoneObsValidator()
+    elif dataset == 'Spectral':
+        return SpectralValidator()
+    elif dataset == 'UmkehrN14':
+        return UmkehrValidator()
+    elif dataset == 'Lidar':
+        return LidarValidator()
+    elif dataset in DATASETS:
+        return DatasetValidator()
+    else:
+        raise ValueError('Invalid dataset {}'.format(dataset))
+
+
+class DatasetValidator:
+    """
+    Superclass for Extended CSV validators of dataset-specific tables.
+    Contains no checks of its own, so all files successfully validate.
+
+    Is directly useful (without subclassing) for datasets that have no
+    errors tied to their tables, and so never have dataset-specific errors.
+    """
+
+    def __init__(self):
+        self.errors = []
+        self.warnings = []
+
+    def _warning(self, error_code, line, message=None):
+        """
+        Record <message> as an error with code <error_code> that took place
+        at line <line> in the input file.
+        """
+
+        LOGGER.warning(message)
+        self.warnings.append((error_code, message, line))
+
+    def _error(self, error_code, line, message=None):
+        """
+        Record <message> as an error with code <error_code> that took place
+        at line <line> in the input file.
+        """
+
+        LOGGER.error(message)
+        self.errors.append((error_code, message, line))
+
+    def check_all(self, extcsv):
+        """
+        Assess any dataset-specific tables inside <extcsv> for errors.
+        Returns True iff no errors were encountered.
+
+        :param extcsv: A parsed Extended CSV file of the appropriate dataset.
+        :returns: True iff the file's dataset-specific tables are error-free.
+        """
+
+        return True
+
+
+class TotalOzoneValidator(DatasetValidator):
```

Comment:
Do subclasses need to run `DatasetValidator.__init__(self)` ?

Reply:
Hmm, guess they should. Thanks for pointing that out, I had thought it was automatically called like in Java.

Reply:
Changed to explicit `__init__` in each class.